lib/pillar/ecto/query.ex

```diff
@@ -0,0 +1,150 @@
+defmodule Pillar.Ecto.Query do
+  alias Pillar.Ecto.QueryBuilder
+
+  @type t :: %__MODULE__{
+          name: iodata,
+          param_count: integer,
+          params: iodata | nil,
+          columns: [String.t()] | nil
+        }
+
+  defstruct name: nil,
+            statement: "",
+            type: :select,
+            params: [],
+            param_count: 0,
+            columns: []
+
+  def new(statement) do
+    %__MODULE__{statement: statement}
+    |> DBConnection.Query.parse([])
+  end
+
+  @spec all(query :: Ecto.Query.t()) :: String.t()
+  def all(query) do
+    sources = QueryBuilder.create_names(query)
+
+    # We now extract all parameters in the query
+    # this should be in the same order as we join the query down below together
+
+    all_params =
+      [
+        {:where, query.wheres},
+        {:limit, query.limit}
+      ]
+      |> Enum.reduce([], fn {for_field, elem}, acc ->
+        List.wrap(elem)
+        |> Enum.reduce(acc, fn expr, acc ->
+          QueryBuilder.param_extractor(for_field, expr, sources, acc)
+        end)
+      end)
+      |> List.flatten()
+      |> Enum.reverse()
```

Comment:
👋 @Zarathustra2 

I wonder if you considered using [`execute/5`](https://hexdocs.pm/ecto/Ecto.Adapter.Queryable.html#c:execute/5) callback to build the query instead of of [`prepare/5`](https://hexdocs.pm/ecto/Ecto.Adapter.Queryable.html#c:prepare/2) to avoid reconstructing params? That's how [`plausible/chto`](https://github.com/plausible/chto/blob/141f09f40d5cb59809f164804a2845e8b35b6e54/lib/ecto/adapters/clickhouse.ex#L197-L224) ended up working.

Reply:
I did not, I didn't even know plausible/chto existed :O

Reply:
Gonna try chto now, maybe I can drop/close my work here

Reply:
@ruslandoga was chto always public or has it been made public recently? I am wondering how I missed it

Reply:
It's always been public in https://github.com/ruslandoga/chto and moved under Plausible org recently. However "always" int his case is just a few months.

Reply:
damn, how did it never come up for me, arg.

Just tested it briefly and seems to work with out issues so far for me. (I am currently struggling with some fragments implementations in this PR). 

I am not sure how my work would compare to your library, feels like your stuff is way more reliable & major than what I have done so far. (I have never worked with Ecto under the hood prior to this PR)

Reply:
> way more reliable & major

The only known user so far is Plausible. So I'm not sure if it's indeed reliable for use cases outside of https://github.com/plausible/analytics ...

It would be great to have more users :)

Reply:
> It would be great to have more users :)

I'm willing to use it. It would be nice if we have a place to chat, maybe on ClickHouse's Slack. I've seen @Zarathustra2 there - where I've found this  MR - but I'm not sure if you are there too, @ruslandoga.